lvsfunc/source.py

```diff
@@ -0,0 +1,206 @@
+
+    if force_lsmas:
+        clip = core.lsmas.LWLibavSource(file, **index_args)
+
+    elif mpls:
+        mpls_in = core.mpls.Read(file, mpls_playlist, mpls_angle)
+        clip = core.std.Splice([core.lsmas.LWLibavSource(mpls_in['clip'][i], **index_args)
+                                for i in range(mpls_in['count'])])
+
+    elif file.endswith('.d2v'):
+        clip = core.d2v.Source(file, **index_args)
+    elif file.endswith('.dgi'):
+        clip = core.dgdecodenv.DGSource(file, **index_args)
+    elif is_image(file):
+        clip = core.imwri.Read(file, **index_args)
+    else:
+        if file.endswith('.m2ts'):
+            clip = core.lsmas.LWLibavSource(file, **index_args)
+        else:
+            clip = core.ffms2.Source(file, **index_args)
+
+    if ref:
+        if ref.format is None:
+            raise ValueError("source: 'Variable-format clips not supported.'")
+        clip = core.std.AssumeFPS(clip, fpsnum=ref.fps.numerator, fpsden=ref.fps.denominator)
+        clip = core.resize.Bicubic(clip, width=ref.width, height=ref.height,
+                                   format=ref.format.id, matrix=get_matrix(ref))
+        if is_image(file):
+            clip = clip * (ref.num_frames - 1)
+
+    return clip
+
+
+class DVDIndexer(ABC):
+    """Abstract DVD indexer interface."""
+    path: Union[Path, str]
+    vps_indexer: Callable[..., vs.VideoNode]
+    ext: str
+
+    def __init__(self, path: Union[Path, str], vps_indexer: Callable[..., vs.VideoNode], ext: str) -> None:
+        self.path = path
+        self.vps_indexer = vps_indexer  # type: ignore
+        self.ext = ext
+        super().__init__()
+
+    @abstractmethod
+    def get_cmd(self, files: List[Path], output: Path) -> List[Any]:
+        """Returns the indexer command"""
+        raise NotImplementedError
+
+
+class D2VWitch(DVDIndexer):
+    """Built-in d2vwitch indexer"""
+    def __init__(self, path: Union[Path, str] = 'd2vwitch',
+                 vps_indexer: Callable[..., vs.VideoNode] = core.d2v.Source, ext: str = '.d2v') -> None:
+        super().__init__(path, vps_indexer, ext)
+
+    def get_cmd(self, files: List[Path], output: Path) -> List[Any]:
+        return [self.path, '--output', output, *files]
+
+
+class DGIndexNV(DVDIndexer):
+    """Built-in DGIndexNV indexer"""
+    def __init__(self, path: Union[Path, str] = 'DGIndexNV',
+                 vps_indexer: Callable[..., vs.VideoNode] = core.dgdecodenv.DGSource, ext: str = '.dgi') -> None:
+        super().__init__(path, vps_indexer, ext)
+
+    def get_cmd(self, files: List[Path], output: Path) -> List[Any]:
+        return [self.path, '-i', ','.join(map(str, files)), '-o', output, '-h']
+
+
+def dvd_source(vob_folder: Union[Path, str], idx: DVDIndexer = D2VWitch(), ifo_file: Optional[Union[Path, str]] = None,
+               extra: bool = False, **kwargs: Any) -> Tuple[List[vs.VideoNode], List[List[int]]]:
```

Comment:
Would like to see a param that returns the entire clip to make this a bit easier to use for newer people. Would just concat every clip I imagine, and return that instead of a list[clips] and list[list[frames]]. Probably a good idea to have that as a default as well, and allow it to be set as False so advanced users can make full use of what this function offers?

Reply:
I feel like there is no point to suse `dvd_source` if you just return the full clip.
Is `std.Splice` that hard to use for the user?